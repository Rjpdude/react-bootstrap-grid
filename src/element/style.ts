import { mapPropertyStyle } from './util'

export type BaseFlexVal<T> = T | 'start' | 'end' | 'center'

export type GridElementDisplayProperty =
  | 'none'
  | 'inline'
  | 'block'
  | 'inline-block'
  | 'table'
  | 'table-row'
  | 'table-cell'
  | 'flex'
  | 'inline-flex'

export const display = mapPropertyStyle<'display'>(undefined, [
  'none',
  'inline',
  'block',
  'inline-block',
  'table',
  'table-row',
  'table-cell',
  'flex',
  'inline-flex'
])

export type GridElementDirectionProperty =
  | 'row'
  | 'row-reverse'
  | 'column'
  | 'column-reverse'

export const direction = mapPropertyStyle<'direction'>('flexDirection', [
  'row',
  'row-reverse',
  'column',
  'column-reverse'
])

export type GridElementJustifyContentProperty = BaseFlexVal<
  'between' | 'around' | 'evenly'
>

export const justifyContent = mapPropertyStyle<'justifyContent'>(
  (val) => ({
    justifyContent:
      val === 'center'
        ? 'center'
        : val === 'start' || val === 'end'
        ? `flex-${val}`
        : `space-${val}`
  }),
  ['start', 'end', 'center', 'between', 'around', 'evenly']
)

export type GridElementAlignContentProperty = BaseFlexVal<
  'between' | 'around' | 'stretch'
>

export const alignContent = mapPropertyStyle<'alignContent'>(
  (val) => ({
    alignContent:
      val === 'start' || val === 'end'
        ? `flex-${val}`
        : val === 'between' || val === 'around'
        ? `space-${val}`
        : val
  }),
  ['start', 'end', 'center', 'between', 'around', 'stretch']
)

export type GridElementAlignItemsProperty = BaseFlexVal<'baseline' | 'stretch'>

export const alignItems = mapPropertyStyle<'alignItems'>(
  (val) => ({
    alignItems: val === 'start' || val === 'end' ? `flex-${val}` : val
  }),
  ['start', 'end', 'center', 'baseline', 'stretch']
)

export type GridElementAlignSelfProperty = BaseFlexVal<
  'auto' | 'baseline' | 'stretch'
>

export const alignSelf = mapPropertyStyle<'alignSelf'>(
  (val) => ({
    alignSelf: val === 'start' || val === 'end' ? `flex-${val}` : val
  }),
  ['start', 'end', 'center', 'auto', 'baseline', 'stretch']
)

export type GridElementOrderProperty =
  | 'first'
  | 'last'
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12

export const order = mapPropertyStyle<'order'>(
  (val) => ({
    order: val === 'first' ? -1 : val === 'last' ? 13 : val
  }),
  ['first', 'last', 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
)

export type GridElementOffsetProperty =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11

export const offset = mapPropertyStyle<'offset'>(
  (val) => ({
    marginLeft: (100 / 12) * val
  }),
  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
)

export type GridElementWrapProperty = boolean | 'reverse'

export const wrap = mapPropertyStyle<'wrap'>(
  (val) => ({
    flexWrap: val === true ? 'wrap' : val === false ? 'nowrap' : 'wrap-reverse'
  }),
  [true, false, 'reverse']
)

export type GridElementGrowthProperty = 0 | 1

export const grow = mapPropertyStyle<'grow'>('flexGrow', [0, 1])

export const shrink = mapPropertyStyle<'shrink'>('flexShrink', [0, 1])

export const fill = mapPropertyStyle<'fill'>(
  (val) =>
    val === true && {
      flex: '1 1 auto'
    },
  [true, false]
)
