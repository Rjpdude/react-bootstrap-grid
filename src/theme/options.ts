export interface ThemeOptions {
  /**
   * The `margin` and `padding` in pixels for all grid elements.
   *
   * @default 15
   */
  spacing: number
  /**
   * The maximum width in pixels of the small breakpoint container.
   *
   * @default 576
   */
  sm: number
  /**
   * The width in pixels to assign to the small breakpoint container.
   *
   * @default 540
   */
  smWidth: number
  /**
   * The maximum width in pixels of the medium breakpoint container.
   *
   * @default 768
   */
  md: number
  /**
   * The width in pixels to assign to the medium breakpoint container.
   *
   * @default 720
   */
  mdWidth: number
  /**
   * The maximum width in pixels of the large breakpoint container.
   *
   * @default 992
   */
  lg: number
  /**
   * The width in pixels to assign to the large breakpoint container.
   *
   * @default 960
   */
  lgWidth: number
  /**
   * The maximum width in pixels of the extra-large breakpoint container.
   *
   * @default 1200
   */
  xl: number
  /**
   * The width in pixels to assign to the extra-large breakpoint container.
   *
   * @default 1140
   */
  xlWidth: number
}

export function getDefaultTheme(): ThemeOptions {
  return {
    spacing: 15,
    sm: 576,
    smWidth: 540,
    md: 768,
    mdWidth: 720,
    lg: 992,
    lgWidth: 960,
    xl: 1200,
    xlWidth: 1140
  }
}
