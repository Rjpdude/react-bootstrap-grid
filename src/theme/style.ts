import { css, createGlobalStyle } from 'styled-components'

export const Global = css`
  html {
    box-sizing: border-box;
    -ms-overflow-style: scrollbar;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }
`

export const GridThemeStyle = createGlobalStyle`
  ${Global}
`
