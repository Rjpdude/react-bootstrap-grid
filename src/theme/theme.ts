import { createElement, createContext, PropsWithChildren } from 'react'
import { ThemeOptions, getDefaultTheme } from './options'
import { GridThemeStyle } from './style'

export const gridThemeContext = createContext(getDefaultTheme())

export function GridTheme(props: PropsWithChildren<Partial<ThemeOptions>>) {
  const style = createElement(GridThemeStyle)
  const context = {
    value: getMergedTheme(props)
  }
  return createElement(
    gridThemeContext.Provider,
    context,
    props.themeStyle !== false && style,
    props.children
  )
}

export function getMergedTheme(themeProps: Partial<ThemeOptions>) {
  const theme = getDefaultTheme()

  Object.keys(theme).forEach((key) => {
    if (themeProps[key]) {
      Object.assign(theme, {
        [key]: themeProps[key]
      })
    }
  })

  return theme
}
