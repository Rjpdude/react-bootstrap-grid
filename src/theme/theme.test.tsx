import React from 'react'
import { mount, shallow } from 'enzyme'
import { getDefaultTheme } from './options'
import { GridThemeStyle } from './style'
import * as Theme from './theme'

describe('Theme', () => {
  describe('Context obj', () => {
    test('Exports context obj', () => {
      expect(Theme.gridThemeContext).toMatchSnapshot()
    })
  })

  describe('GridTheme component', () => {
    test('Exports grid theme component', () => {
      expect(Theme.GridTheme).toEqual(expect.any(Function))
    })

    test('Renders theme style', () => {
      expect(
        shallow(<Theme.GridTheme />)
          .find(GridThemeStyle)
          .exists()
      ).toBe(true)
    })

    test('Omits theme style when `themeStyle` prop is false', () => {
      expect(
        shallow(<Theme.GridTheme themeStyle={false} />)
          .find(GridThemeStyle)
          .exists()
      ).toBe(false)
    })

    test('Renders children', () => {
      const gridTheme = mount(
        <Theme.GridTheme>
          <p>Child element</p>
        </Theme.GridTheme>
      )
      expect(gridTheme.find('p').text()).toBe('Child element')
    })

    test('Supplies theme context', () => {
      const Child = () => (
        <p>{React.useContext(Theme.gridThemeContext).spacing}</p>
      )

      const theme = mount(
        <Theme.GridTheme spacing={16}>
          <Child />
        </Theme.GridTheme>
      )

      expect(theme.text()).toBe('16')
    })
  })

  describe('getMergedTheme', () => {
    test('Supplies default theme object on empty props', () => {
      expect(Theme.getMergedTheme({})).toEqual(getDefaultTheme())
    })

    test('Omits invalid theme props', () => {
      // @ts-ignore
      expect(Theme.getMergedTheme({ children: null, someProp: 5 })).toEqual(
        getDefaultTheme()
      )
    })

    test('Merges supplied theme options', () => {
      const options = { spacing: 18, xlWidth: 1200 }
      const defaultTheme = getDefaultTheme()
      const merged = Theme.getMergedTheme(options)

      Object.entries(options).forEach(([key, value]) => {
        expect(merged[key]).toBe(value)
      })

      Object.entries(merged).forEach(([key, value]) => {
        if (!options[key]) {
          expect(value).toEqual(defaultTheme[key])
        }
      })
    })
  })
})
