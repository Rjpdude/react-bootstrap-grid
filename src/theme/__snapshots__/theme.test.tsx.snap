// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context obj Exports context obj 1`] = `
Object {
  "$$typeof": Symbol(react.context),
  "Consumer": Object {
    "$$typeof": Symbol(react.context),
    "_calculateChangedBits": null,
    "_context": [Circular],
  },
  "Provider": Object {
    "$$typeof": Symbol(react.provider),
    "_context": [Circular],
  },
  "_calculateChangedBits": null,
  "_currentRenderer": null,
  "_currentRenderer2": null,
  "_currentValue": Object {
    "lg": 992,
    "lgWidth": 960,
    "md": 768,
    "mdWidth": 720,
    "sm": 576,
    "smWidth": 540,
    "spacing": 15,
    "themeStyle": true,
    "xl": 1200,
    "xlWidth": 1140,
  },
  "_currentValue2": Object {
    "lg": 992,
    "lgWidth": 960,
    "md": 768,
    "mdWidth": 720,
    "sm": 576,
    "smWidth": 540,
    "spacing": 15,
    "themeStyle": true,
    "xl": 1200,
    "xlWidth": 1140,
  },
  "_threadCount": 0,
}
`;
